File: LeetCode/42.Trapping_Rain_Water/42.Trapping_Rain_Water.java
class Solution {
    public int trap(int[] height) {
        // init monotonic stack
        int n = height.length;
        Stack<int[]> stack = new Stack<>();

        int trapArea = 0;
        for (int i = 0; i < n; i++) {
            int h = height[i];
            while (!stack.isEmpty() && stack.peek()[1] < h) {
                int[] last = stack.pop();
                if (stack.isEmpty())
                    break;
                int[] top = stack.peek();
                int dis = (i - top[0] - 1);
                int heightSide = Math.min(top[1], h) - last[1];
                trapArea += (heightSide * dis);
            }

            stack.push(new int[] { i, h });
        }

        return trapArea;
    }
}

File: LeetCode/42.Trapping_Rain_Water/42.Trapping_Rain_Water.swift
class Solution {
    func trap(_ height: [Int]) -> Int {
        // monotonic stack decreasing order
        var stack: [Int] = []
        var res = 0

        for (i, h) in height.enumerated() {
            while let last = stack.last, height[last] < h {
                let bottom = height[stack.removeLast()]
                if let left = stack.last {
                    let width = i - left - 1
                    let curHeight = min(h, height[left]) - bottom
                    let area = width * curHeight
                    res += area
                }
            }

            stack.append(i)
        }

        return res
    }
}
